Add Ubiquitous Language reference (#71)
MinBZK/ai-validation: docs/Way-of-Working/UbiquitousLanguage.md

# Ubiquitous Language

> For clarity and consistency, this document defines key terms used within our team across various domains (e.g., Data
> Science, IT).

Reviewer: ```suggestion
For clarity and consistency, this document defines key terms used within our team across various domains (e.g., 
Machine Learning, Data Science, IT).
```

The goal is to provide a list of commonly used terms and terms that are good to know. It is not to a full reference of
all existing terms.

> Terms can and have been used 'as is' from sources mentioned at the end of this document.

Reviewer: I'm not sure it is helpful to duplicate data from elsewhere (can't we just link then?). This list seems a little long. Can't we limit it to language we actually encouter/use?

Reviewer 2: I second the remark of Anne. I think it would be a good idea to only add terms to this list that we encounter, otherwise the items on the list seem a bit random.

## Terms used in Data Science

### Algorithm

> An algorithm is a sequence of repeatable steps, often expressed mathematically, written by a human and executed by a
> computer, to solve a certain type of data science problem. The main concept is that an algorithm takes in some input and

Reviewer: ```suggestion
An algorithm is a sequence of repeatable steps, often expressed mathematically, written by a human and executed by a
a human or computer, to solve a certain type of data science problem. The main concept is that an algorithm takes in some input and
```

Reviewer 2: I would not limit the definition of an algorithm to "solve certain type of data science problems". The common definition of an algorithm is broader than this. I would take any definition from a standard textbook, for example the definition [here](https://ada-developers-academy.github.io/jumpstart-textbook/jumpstart-lessons/problem-solving/algorithms/#:~:text=In%20The%20Art%20of%20Computer,a%20reasonable%20amount%20of%20time) of Knuth coming from his classic textbook.

produces an output, and the same input will always produce the same output. In machine learning, algorithms take in
input in the form of data and [hyperparameters](https://www.datacamp.com/blog/data-science-glossary#hyperparameter),
identify and learn common patterns from the data, and produce outputs in the form of predictions.

### Bias

> Bias refers to the tendency of models to underfit the data, leading to inaccurate predictions in machine learning and
> data science. Moreover, bias can also mean algorithmic bias, which refers to the propensity of machine learning models
> to reproduce harmful societal biases by treating different groups of individuals differently based on protected
> attributes such as race, sexual orientation, gender identify, age, pregnancy and more.

Reviewer: The tendency of a model to favor certain outcomes over others, potentially leading to inaccuracy or unfairness.
```suggestion
attributes such as race, sexual orientation, gender identify, age, pregnancy and more.
```
```suggestion
attributes such as race, sexual orientation, gender identify, age, pregnancy and more.
```

> Bias refers to the tendency of models to underfit the data, leading to inaccurate predictions in machine learning and
> data science. Moreover, bias can also mean algorithmic bias, which refers to the propensity of machine learning models
> to reproduce harmful societal biases by treating different groups of individuals differently based on protected
> attributes such as race, sexual orientation, gender identify, age, pregnancy and more.

Reviewer: I don't think this is entirely correct. Bias and underfitting are conflated.

```suggestion
Bias refers to the tendency of a model to favor certain outcomes over others, potentially leading to inaccuracy or unfairness.

Moreover, bias can also mean algorithmic bias, which refers to the propensity of machine learning models
to reproduce harmful societal biases by treating different groups of individuals differently based on protected
attributes such as race, sexual orientation, gender identify, age, pregnancy and more.
```

> Bias refers to the tendency of models to underfit the data, leading to inaccurate predictions in machine learning and
> data science. Moreover, bias can also mean algorithmic bias, which refers to the propensity of machine learning models
> to reproduce harmful societal biases by treating different groups of individuals differently based on protected
> attributes such as race, sexual orientation, gender identify, age, pregnancy and more.
> 

Reviewer: I don't think this is entirely correct. Bias and underfitting are conflated.

```suggestion
Bias refers to the tendency of a model to favor certain outcomes over others, potentially leading to inaccuracy or unfairness.

Moreover, bias can also mean algorithmic bias, which refers to the propensity of machine learning models
to reproduce harmful societal biases by treating different groups of individuals differently based on protected
attributes such as race, sexual orientation, gender identify, age, pregnancy and more.
```

Reviewer 2: Bias can mean many different things and often very precise things, for example [statistical bias](https://en.wikipedia.org/wiki/Bias_of_an_estimator). I would not mention these definitions here, but just focus on algorithmic bias to avoid confusion.

> ### Big Data

Reviewer: do we need this?

Big Data is the field that revolves around processing, treating, and extracting information
from [data sets](https://en.wikipedia.org/wiki/Data_set) that are too large or complex for traditional data processing
tools. Big data is defined by the five Vs; velocity: the speed of data generation; volume: the amount of data generated;
variety: the variety of data types e.g., text, images, tabular data etc.; veracity: the quality and truthfulness of the
data; and value: the data’s propensity to be translated to valuable business insights.

### Classification

Classification is a supervised learning problem when it is necessary to predict categorical outcomes based on input
features. Examples of classification problems are fraud detection (e.g., is this transaction fraudulent given the set of
input features?) and email spam filters (e.g. Is this email spam or not?). Commonly used classification algorithms are

> k-nearest neighbors, decision trees, random forest, etc.

Reviewer 2: ```suggestion
k-nearest neighbors, decision trees, random forest, etc. Another common use case of classification is classification of images (e.g. does this image contain a human or not); commonly used algorithms are convolutional neural networks. 
```

### Clustering

Clustering is an unsupervised learning problem concerned with grouping all the observations of a dataset according to
their similarity by some common characteristics. Unlike a classification problem, these groups (called clusters) are not
predefined by humans, but identified by machine learning algorithms while learning the input data. The elements in each
cluster are similar among themselves and different from all the others. Common clustering algorithms are k-means,
hierarchical clustering, spectral clustering, etc..

### Confusion Matrix

> A confusion matrix is a table illustrating the predictive performance of a classification model. Usually, a confusion
> matrix is created for a binary output (i.e., prediction problems with only two types of prediction—e.g., whether a
> transaction is fraudulent or not), so the resulting table is a two by two table. A confusion matrix represents the

Reviewer: This isn't correct, doesn't need to be binary. 
```suggestion
A confusion matrix is a table illustrating the predictive performance of a classification model. Usually, a confusion
matrix is created for a classification model. A confusion matrix represents the
```

### Covariance

Covariance is a measure of the relationship between two variables.
Unlike [variance](https://www.datacamp.com/blog/data-science-glossary#variance) that measures variations inside the same
variable, covariance shows how variations in one variable influence changes in the second one. Covariance is used for
calculating a correlation coefficient.

### Cross-Validation

Cross-validation is a resampling method when training machine learning models that splits labeled data into training and
test sets. In each iteration of cross validation, different parts of the data are used to train and test the model. The
training set is used to train a model, and the test set is used to make predictions and compare them with the actual
labels for those entries. Afterwards, an overall accuracy metric is calculated to estimate the predictive performance of
the resulting model.

### Data Enrichment

Data enrichment is the processes of enhancing, refining, and augmenting raw data, to make it more useful for the
organization and, as a result, to obtain more meaningful business insights and optimize predictive analytics.

### Dataframe

> A dataframe is the tabular data structure with labeled axes (rows and columns) of potentially different types.

Reviewer 2: I think the term dataframe is specific to the Pandas package. Maybe good to mention that here.

### Data Mining

Data mining is the process of collecting relevant data from various sources, cleaning and transforming it to the right
format, detecting and extracting meaningful hidden trends, patterns, and interconnections between the data, and
communicating actionable insights to help the organization make data-driven decisions and develop better strategies. For
this purpose, various analytical and modeling techniques are used, including statistical analysis, data visualization,
regression and classification.

### Dataset

A dataset is a collection of data of one or many types representing real-life or synthetically generated observations,
and used for statistical analysis or data modeling. The data of a dataset can be gathered from many sources and is
typically stored in some kind of data structure, most commonly a table, where the columns correspond to different
variables, and the rows to different data entries.

### Deep Learning (DL)

Deep learning is a subset of machine learning algorithms based on

> multilayered [artificial neural networks (ANN)](https://www.datacamp.com/blog/data-science-glossary#artificial-neural-networks-(ann))

Reviewer 2: ```suggestion
multilayered [artificial neural networks (ANN)](https://www.datacamp.com/blog/data-science-glossary#artificial-neural-networks-(ann)), commonly known as deep neural networks, 
```

> that are largely inspired by the structure of the brain. ANN are very flexible and can learn from huge amounts of data,

Reviewer 2: ```suggestion
that are inspired by the structure of the brain. ANN are very flexible and can learn from huge amounts of data,
```

The neural analology is not as strong as people tend to think.

### Logistic Regression

Logistic regression is a regression algorithm that uses a logistic function on the input features to predict the class
probability or directly the class label for the target variable. In the second case, the output represents a set of
categories instead of continuous values, meaning that the logistic regression acts here as a classification technique. A
typical data science use case for logistic regression is predicting the likelihood of customer churn.

### Normalization

Normalization is the process of rescaling the data so that all the attributes have the same scale. Normalization is
necessary for making a meaningful comparison between the attributes and also is required for some machine learning
algorithms.

### Overfitting

Overfitting refers to when a model learns too much information from the training set including potential noise and
outliers. As a result, it becomes too complex, too conditioned on the particular training set, and fails to adequately
perform on unseen data. Overfitting leads to high variance on
the [bias-variance tradeoff](https://www.datacamp.com/blog/data-science-glossary#bias-variance-tradeoff).

> ### Precision

Reviewer 2: If you want to mention precision maybe also mention related metrics such as recall and F1 scores

regression and ridge regression. A typical regression problem is price prediction.

### Reinforcement Learning (RL)

Reinforcement learning (RL) is a stand-alone branch of machine learning (neither supervised nor unsupervised) where an
algorithm gradually learns by interacting with an environment. RL makes decisions based on its past experience about
which actions can bring it closer to a stated goal. By receiving rewards for correct actions and penalties for wrong
ones, the algorithm finds out the optimal strategy to maximize its performance. Examples of RL algorithms include
game-playing machine learning systems such as chess engines and video game agents.

### Supervised Learning

Supervised learning is a branch of machine learning concerned with teaching a model on a labeled training set of
historical data. Supervised learning learns the relationship between inputs and outputs and then measures how accurately
it predicts the outputs for a test set with the known actual outputs. In this way it can be used later to make
predictions on completely new data. Supervised learning algorithms include linear and logistic regressions, decision

> trees, and SVM. Examples of common tasks include predicting house prices and classifying messages as spam or ham.

Reviewer 2: I would definitely mention image classification here as well.

# Ubiquitous Language

For clarity and consistency, this document defines some terms used within our team where the meaning in Data Science or
Computer Science differs, and terms that are for any reason good to mention.

For a full reference for Machine Learning, we
recommend [ML Fundamentals from Google](https://developers.google.com/machine-learning/glossary/fundamentals).

> 
> ## Terms

Reviewer 3: For the time being, shall we remove this header?

```suggestion

## Terms
```
